src/client/Microsoft.Identity.Client/Kerberos/KerberosSupplementalTicketManager.cs

```diff
@@ -0,0 +1,219 @@
+        /// <remarks>
+        /// Can throws <see cref="Win32Exception"/> if error occurs while searching ticket information from Ticket Cache.
+        /// </remarks>
+        public static byte[] GetKerberosTicketFromWindowsTicketCache(string servicePrincipalName)
+        {
+            return GetKerberosTicketFromWindowsTicketCache(servicePrincipalName, _defaultLogonId);
+        }
+
+        /// <summary>
+        /// Reads a Kerberos Service Ticket associated with given service principal name from
+        /// current user's Ticket Cache.
+        /// </summary>
+        /// <param name="servicePrincipalName">Service principal name to find associated Kerberos Ticket.</param>
+        /// <param name="logonId">The Logon Id of the user owning the ticket cache.
+        /// The default of 0 represents the currently logged on user.</param>
+        /// <returns>Byte stream of searched Kerberos Ticket information if exists. Null, otherwise.</returns>
+        /// <remarks>
+        /// Can throws <see cref="Win32Exception"/> if error occurs while searching ticket information from Ticket Cache.
+        /// </remarks>
+        public static byte[] GetKerberosTicketFromWindowsTicketCache(string servicePrincipalName, long logonId)
+        {
+#if SUPPORT_KERBEROS
+            using (var reader = new Win32.TicketCacheReader(servicePrincipalName, logonId))
+            {
+                return reader.RequestToken();
+            }
+#else
+            return null;
+#endif
+        }
+
+        /// <summary>
+        /// Gets the KRB-CRED Kerberos Ticket information as byte stream.
+        /// </summary>
+        /// <param name="ticket">Kerberos ticket object to save.</param>
+        /// <returns>Byte stream representaion of KRB-CRED Kerberos Ticket if it contains valid ticket information.
+        /// Null, otherwise.</returns>
+        public static byte[] GetKrbCred(KerberosSupplementalTicket ticket)
+        {
+            if (!string.IsNullOrEmpty(ticket.KerberosMessageBuffer))
+            {
+                return Convert.FromBase64String(ticket.KerberosMessageBuffer);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Add Claims to body parameter for POST request.
+        /// </summary>
+        /// <param name="oAuth2Client"><see cref="OAuth2Client"/> object for Token request.</param>
+        /// <param name="requestParams"><see cref="AuthenticationRequestParameters"/> containing request parameters.</param>
+        internal static void AddKerberosTicketClaim(
+            OAuth2Client oAuth2Client,
+            AuthenticationRequestParameters requestParams)
+        {
+            string kerberosClaim = GetKerberosTicketClaim(
+                requestParams.RequestContext.ServiceBundle.Config.KerberosServicePrincipalName,
+                requestParams.RequestContext.ServiceBundle.Config.TicketContainer);
+
+            if (string.IsNullOrEmpty(kerberosClaim))

```

Comment:
In `AcquireTokenSilent`, if claims are defined, then we bypass the token cache. Should we bypass the token cache if the kerberos claim is defined?

Reply:
Yes. Since the Kerberos Ticket can be updated, it's better not to use cached token if possible.